src/device/device_ledger.cpp

```diff
@@ -0,0 +1,2057 @@
+    #define INS_GET_SUBADDRESS_SPEND_PUBLIC_KEY 0x4A
+    #define INS_GET_SUBADDRESS_SECRET_KEY       0x4C
+
+    #define INS_OPEN_TX                         0x70
+    #define INS_SET_SIGNATURE_MODE              0x72
+    #define INS_GET_ADDITIONAL_KEY              0x74
+    #define INS_STEALTH                         0x76
+    #define INS_BLIND                           0x78
+    #define INS_UNBLIND                         0x7A
+    #define INS_VALIDATE                        0x7C
+    #define INS_MLSAG                           0x7E
+    #define INS_CLOSE_TX                        0x80
+
+
+    #define INS_GET_RESPONSE                    0xc0
+
+
+    void DeviceLedger::logCMD() {
+      if (apdu_verbose) {
+        char  strbuffer[1024];
+        sprintf(strbuffer, "%.02x %.02x %.02x %.02x %.02x ",
+          this->buffer_send[0],
+          this->buffer_send[1],
+          this->buffer_send[2],
+          this->buffer_send[3],
+          this->buffer_send[4]
+          );
+        buffer_to_str(strbuffer+strlen(strbuffer), (char*)(this->buffer_send+5), this->length_send-5);
+        MDEBUG( "CMD  :" << strbuffer);
+      }
+    }
+
+    void DeviceLedger::logRESP() {
+      if (apdu_verbose) {
+        char  strbuffer[1024];
+        sprintf(strbuffer, "%.02x%.02x ",
+          this->buffer_recv[this->length_recv-2],
+          this->buffer_recv[this->length_recv-1]
+          );
+        buffer_to_str(strbuffer+strlen(strbuffer), (char*)(this->buffer_recv), this->length_recv-2);
+        MDEBUG( "RESP :" << strbuffer);
+
+      }
+    }
+
+    /* -------------------------------------------------------------- */
+    DeviceLedger::DeviceLedger() : exiting(false) DEVICE_CONTROLE {
+      this->id = device_id++;
+      this->hCard   = 0;
+      this->hContext = 0;
+      this->reset_buffer();
+      if (this->id)
+        MDEBUG( "Device "<<this->id <<" Created");
+    }
+
+    DeviceLedger::DeviceLedger(const DeviceLedger &device): DeviceLedger() {
+      this->set_name(device.name);
+      MDEBUG( "Device "<<device.id <<" cloned. Device "<<this->id <<" created");
+    }
+
+    DeviceLedger& DeviceLedger::operator=(const DeviceLedger &device) {
+      this->set_name(device.name);
+      this->id = device_id++;
+      this->hCard   = 0;
+      this->hContext = 0;
+      this->reset_buffer();
+      this->exiting = device.exiting;
+      MDEBUG( "Device "<<device.id <<" assigned. Device "<<this->id <<" created");
+      return *this;
+    }
+
+    DeviceLedger::~DeviceLedger() {
+      if (this->id == 0) {
+        exiting = true;
+      }
+      this->release();
+      if (this->id) {
+        MDEBUG( "Device "<<this->id <<" Destroyed");
+      }
+    }
+
+    void DeviceLedger::lock_device()   {
+      MDEBUG( "Ask for LOCKING for device "<<this->id);
+      device_locker.lock();
+      MDEBUG( "Device "<<this->id << " LOCKed");
+    }
+    void DeviceLedger::unlock_device() {
+      MDEBUG( "Ask for UNLOCKING for device "<<this->id);
```

Comment:
Don't do this. It can throw, and leave the device locked. Best case this would have to be wrapped in its own try/catch block.

Reply:
Not sure to understand what you suggest. MDEBUG may throw exception? 
So I modified with:

      try {
        MDEBUG( "Ask for UNLOCKING for device "<<this->id);
      } catch (...) {
      }

Reply:
Yes MDEBUG can throw in many different locations internally (it allocates memory, even if debug logging if off) .... yes I know.